services/pipeline/src/test/java/tech/pegasys/pantheon/services/pipeline/BatchingReadPipeTest.java

```diff
@@ -0,0 +1,134 @@
+/*
+ * Copyright 2019 ConsenSys AG.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except in compliance with
+ * the License. You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software distributed under the License is distributed on
+ * an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the License for the
+ * specific language governing permissions and limitations under the License.
+ */
+package tech.pegasys.pantheon.services.pipeline;
+
+import static java.util.Arrays.asList;
+import static org.assertj.core.api.Assertions.assertThat;
+import static org.mockito.Mockito.mock;
+import static org.mockito.Mockito.times;
+import static org.mockito.Mockito.verify;
+import static org.mockito.Mockito.verifyZeroInteractions;
+import static tech.pegasys.pantheon.metrics.noop.NoOpMetricsSystem.NO_OP_COUNTER;
+
+import tech.pegasys.pantheon.metrics.Counter;
+
+import java.util.ArrayList;
+import java.util.List;
+
+import org.junit.Test;
+
+public class BatchingReadPipeTest {
+
+  private final Pipe<String> source = new Pipe<>(10, NO_OP_COUNTER, NO_OP_COUNTER);
+  private final Counter batchCounter = mock(Counter.class);
+  private final BatchingReadPipe<String> batchingPipe =
+      new BatchingReadPipe<>(source, 3, batchCounter);
+
+  @Test
+  public void shouldGetABatchOfAvailableItems() {
+    source.put("a");
+    source.put("b");
+    source.put("c");
+    source.put("d");
+
+    assertThat(batchingPipe.get()).containsExactly("a", "b", "c");
+  }
+
+  @Test
+  public void shouldNotWaitToFillBatch() {
+    source.put("a");
+    assertThat(batchingPipe.get()).containsExactly("a");
+  }
+
+  @Test
+  public void shouldPollForNextBatchWhenAvailable() {
+    source.put("a");
+    assertThat(batchingPipe.poll()).containsExactly("a");
+  }
+
+  @Test
+  public void shouldReturnNullFromPollWhenNoItemsAvailable() {
+    assertThat(batchingPipe.poll()).isNull();
+  }
+
+  @Test
+  public void shouldHaveMoreItemsWhileSourceHasMoreItems() {
+    assertThat(batchingPipe.hasMore()).isTrue();
+
+    source.put("a");
+    source.put("b");
+    source.put("c");
+
+    assertThat(batchingPipe.hasMore()).isTrue();
+
+    source.close();
+
+    assertThat(batchingPipe.hasMore()).isTrue();
+
+    assertThat(batchingPipe.get()).containsExactly("a", "b", "c");
+
+    assertThat(batchingPipe.hasMore()).isFalse();
+
+    assertThat(batchingPipe.get()).isNull();
+    assertThat(batchingPipe.poll()).isNull();
+  }
+
+  @Test
+  public void shouldAddAtMostOneItemWhenDraining() {
+    // Collecting a batch of batches is pretty silly so only ever drain one batch at a time.
```

Comment:
🤔

Reply:
For the record, nothing actually ever calls this but it seems better to give it a vaguely sane implementation than to throw an exception and not implement the underlying interface properly.